{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>3214482</td>\n",
       "      <td>3216968</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Xkr4\"; transcript_id \"NM_001011874\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>3216022</td>\n",
       "      <td>3216024</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Xkr4\"; transcript_id \"NM_001011874\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>3216025</td>\n",
       "      <td>3216968</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"Xkr4\"; transcript_id \"NM_001011874\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>3421702</td>\n",
       "      <td>3421901</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>gene_id \"Xkr4\"; transcript_id \"NM_001011874\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>3421702</td>\n",
       "      <td>3421901</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Xkr4\"; transcript_id \"NM_001011874\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>595681</th>\n",
       "      <td>chrY</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>90793296</td>\n",
       "      <td>90793417</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"Erdr1\"; transcript_id \"NM_133362\"; ge...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>595682</th>\n",
       "      <td>chrY</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>90793296</td>\n",
       "      <td>90793417</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Erdr1\"; transcript_id \"NM_133362\"; ge...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>595683</th>\n",
       "      <td>chrY</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>90816349</td>\n",
       "      <td>90816444</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>gene_id \"Erdr1\"; transcript_id \"NM_133362\"; ge...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>595684</th>\n",
       "      <td>chrY</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>90816349</td>\n",
       "      <td>90816465</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Erdr1\"; transcript_id \"NM_133362\"; ge...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>595685</th>\n",
       "      <td>chrY</td>\n",
       "      <td>unknown</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>90816445</td>\n",
       "      <td>90816447</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Erdr1\"; transcript_id \"NM_133362\"; ge...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>595686 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       seqname   source     feature     start       end score strand frame  \\\n",
       "0         chr1  unknown        exon   3214482   3216968     .      -     .   \n",
       "1         chr1  unknown  stop_codon   3216022   3216024     .      -     .   \n",
       "2         chr1  unknown         CDS   3216025   3216968     .      -     2   \n",
       "3         chr1  unknown         CDS   3421702   3421901     .      -     1   \n",
       "4         chr1  unknown        exon   3421702   3421901     .      -     .   \n",
       "...        ...      ...         ...       ...       ...   ...    ...   ...   \n",
       "595681    chrY  unknown         CDS  90793296  90793417     .      +     0   \n",
       "595682    chrY  unknown        exon  90793296  90793417     .      +     .   \n",
       "595683    chrY  unknown         CDS  90816349  90816444     .      +     1   \n",
       "595684    chrY  unknown        exon  90816349  90816465     .      +     .   \n",
       "595685    chrY  unknown  stop_codon  90816445  90816447     .      +     .   \n",
       "\n",
       "                                                attribute  \n",
       "0       gene_id \"Xkr4\"; transcript_id \"NM_001011874\"; ...  \n",
       "1       gene_id \"Xkr4\"; transcript_id \"NM_001011874\"; ...  \n",
       "2       gene_id \"Xkr4\"; transcript_id \"NM_001011874\"; ...  \n",
       "3       gene_id \"Xkr4\"; transcript_id \"NM_001011874\"; ...  \n",
       "4       gene_id \"Xkr4\"; transcript_id \"NM_001011874\"; ...  \n",
       "...                                                   ...  \n",
       "595681  gene_id \"Erdr1\"; transcript_id \"NM_133362\"; ge...  \n",
       "595682  gene_id \"Erdr1\"; transcript_id \"NM_133362\"; ge...  \n",
       "595683  gene_id \"Erdr1\"; transcript_id \"NM_133362\"; ge...  \n",
       "595684  gene_id \"Erdr1\"; transcript_id \"NM_133362\"; ge...  \n",
       "595685  gene_id \"Erdr1\"; transcript_id \"NM_133362\"; ge...  \n",
       "\n",
       "[595686 rows x 9 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load GTF file - replace with your actual file path\n",
    "gtf_df = pd.read_csv('genes.gtf', sep='\\t', comment='#', header=None,\n",
    "                     names=['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute'])\n",
    "# display(gtf_df.head(20))\n",
    "# features = set(gtf_df['feature'])\n",
    "# display(features)\n",
    "gtf_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>4409185</td>\n",
       "      <td>4409187</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Rp1\"; transcript_id \"NM_001195662\"; g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>4490928</td>\n",
       "      <td>4492668</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Sox17\"; transcript_id \"NM_011441\"; ge...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>4491716</td>\n",
       "      <td>4491718</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Sox17\"; transcript_id \"NM_011441\"; ge...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>4491719</td>\n",
       "      <td>4492668</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"Sox17\"; transcript_id \"NM_011441\"; ge...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>4493100</td>\n",
       "      <td>4493406</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"Sox17\"; transcript_id \"NM_011441\"; ge...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>4493100</td>\n",
       "      <td>4493466</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Sox17\"; transcript_id \"NM_011441\"; ge...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>4493404</td>\n",
       "      <td>4493406</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Sox17\"; transcript_id \"NM_011441\"; ge...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>4493772</td>\n",
       "      <td>4493863</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Sox17\"; transcript_id \"NM_011441\"; ge...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>4495136</td>\n",
       "      <td>4495942</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Sox17\"; transcript_id \"NM_011441\"; ge...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>4496291</td>\n",
       "      <td>4496413</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Sox17\"; transcript_id \"NM_011441\"; ge...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>4773200</td>\n",
       "      <td>4774516</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Mrpl15\"; transcript_id \"NR_033530\"; g...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   seqname   source      feature    start      end score strand frame  \\\n",
       "26    chr1  unknown  start_codon  4409185  4409187     .      -     .   \n",
       "27    chr1  unknown         exon  4490928  4492668     .      -     .   \n",
       "28    chr1  unknown   stop_codon  4491716  4491718     .      -     .   \n",
       "29    chr1  unknown          CDS  4491719  4492668     .      -     2   \n",
       "30    chr1  unknown          CDS  4493100  4493406     .      -     0   \n",
       "31    chr1  unknown         exon  4493100  4493466     .      -     .   \n",
       "32    chr1  unknown  start_codon  4493404  4493406     .      -     .   \n",
       "33    chr1  unknown         exon  4493772  4493863     .      -     .   \n",
       "34    chr1  unknown         exon  4495136  4495942     .      -     .   \n",
       "35    chr1  unknown         exon  4496291  4496413     .      -     .   \n",
       "36    chr1  unknown         exon  4773200  4774516     .      -     .   \n",
       "\n",
       "                                            attribute  \n",
       "26  gene_id \"Rp1\"; transcript_id \"NM_001195662\"; g...  \n",
       "27  gene_id \"Sox17\"; transcript_id \"NM_011441\"; ge...  \n",
       "28  gene_id \"Sox17\"; transcript_id \"NM_011441\"; ge...  \n",
       "29  gene_id \"Sox17\"; transcript_id \"NM_011441\"; ge...  \n",
       "30  gene_id \"Sox17\"; transcript_id \"NM_011441\"; ge...  \n",
       "31  gene_id \"Sox17\"; transcript_id \"NM_011441\"; ge...  \n",
       "32  gene_id \"Sox17\"; transcript_id \"NM_011441\"; ge...  \n",
       "33  gene_id \"Sox17\"; transcript_id \"NM_011441\"; ge...  \n",
       "34  gene_id \"Sox17\"; transcript_id \"NM_011441\"; ge...  \n",
       "35  gene_id \"Sox17\"; transcript_id \"NM_011441\"; ge...  \n",
       "36  gene_id \"Mrpl15\"; transcript_id \"NR_033530\"; g...  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf_df.iloc[26:37]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>3214482</td>\n",
       "      <td>3216968</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Xkr4\"; transcript_id \"NM_001011874\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>3421702</td>\n",
       "      <td>3421901</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Xkr4\"; transcript_id \"NM_001011874\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>3670552</td>\n",
       "      <td>3671498</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Xkr4\"; transcript_id \"NM_001011874\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>4290846</td>\n",
       "      <td>4293012</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Rp1\"; transcript_id \"NM_001195662\"; g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>4343507</td>\n",
       "      <td>4350091</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Rp1\"; transcript_id \"NM_011283\"; gene...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>595672</th>\n",
       "      <td>chrY</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>83927454</td>\n",
       "      <td>83927911</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"LOC100040911\"; transcript_id \"NM_0011...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>595674</th>\n",
       "      <td>chrY</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>85528525</td>\n",
       "      <td>85529519</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"LOC100040911\"; transcript_id \"NM_0011...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>595678</th>\n",
       "      <td>chrY</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>90785442</td>\n",
       "      <td>90785979</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Erdr1\"; transcript_id \"NM_133362\"; ge...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>595682</th>\n",
       "      <td>chrY</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>90793296</td>\n",
       "      <td>90793417</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Erdr1\"; transcript_id \"NM_133362\"; ge...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>595684</th>\n",
       "      <td>chrY</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>90816349</td>\n",
       "      <td>90816465</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Erdr1\"; transcript_id \"NM_133362\"; ge...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>284820 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       seqname   source feature     start       end score strand frame  \\\n",
       "0         chr1  unknown    exon   3214482   3216968     .      -     .   \n",
       "4         chr1  unknown    exon   3421702   3421901     .      -     .   \n",
       "6         chr1  unknown    exon   3670552   3671498     .      -     .   \n",
       "8         chr1  unknown    exon   4290846   4293012     .      -     .   \n",
       "11        chr1  unknown    exon   4343507   4350091     .      -     .   \n",
       "...        ...      ...     ...       ...       ...   ...    ...   ...   \n",
       "595672    chrY  unknown    exon  83927454  83927911     .      -     .   \n",
       "595674    chrY  unknown    exon  85528525  85529519     .      -     .   \n",
       "595678    chrY  unknown    exon  90785442  90785979     .      +     .   \n",
       "595682    chrY  unknown    exon  90793296  90793417     .      +     .   \n",
       "595684    chrY  unknown    exon  90816349  90816465     .      +     .   \n",
       "\n",
       "                                                attribute  \n",
       "0       gene_id \"Xkr4\"; transcript_id \"NM_001011874\"; ...  \n",
       "4       gene_id \"Xkr4\"; transcript_id \"NM_001011874\"; ...  \n",
       "6       gene_id \"Xkr4\"; transcript_id \"NM_001011874\"; ...  \n",
       "8       gene_id \"Rp1\"; transcript_id \"NM_001195662\"; g...  \n",
       "11      gene_id \"Rp1\"; transcript_id \"NM_011283\"; gene...  \n",
       "...                                                   ...  \n",
       "595672  gene_id \"LOC100040911\"; transcript_id \"NM_0011...  \n",
       "595674  gene_id \"LOC100040911\"; transcript_id \"NM_0011...  \n",
       "595678  gene_id \"Erdr1\"; transcript_id \"NM_133362\"; ge...  \n",
       "595682  gene_id \"Erdr1\"; transcript_id \"NM_133362\"; ge...  \n",
       "595684  gene_id \"Erdr1\"; transcript_id \"NM_133362\"; ge...  \n",
       "\n",
       "[284820 rows x 9 columns]"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter for transcripts to get 'feature' == 'exons'\n",
    "tss_df = gtf_df[gtf_df['feature'] == 'exon']\n",
    "\n",
    "tss_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to parse the attributes column and return a dictionary\n",
    "def parse_attributes(attributes_str):\n",
    "    # Split the string by semicolon and space, then by space within each pair\n",
    "    # Assuming that the GTF format is consistent with the gene_name being quoted\n",
    "    return {attr.strip().split(' ')[0]: attr.strip().split(' ')[1].replace('\"', '')\n",
    "            for attr in attributes_str.strip().split(';') if attr}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "23337\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/z7/2__sng712j3bvbdc64_2jnpr0000gn/T/ipykernel_80685/1137306480.py:13: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  tss_df[['gene_name']] = tss_df.apply(extract_gene_name, axis=1)\n"
     ]
    }
   ],
   "source": [
    "# Define a function to extract gene name and TSS based on the strand\n",
    "def extract_tss_info(row):\n",
    "    tss_start = row['start'] if row['strand'] == '+' else row['end']\n",
    "    return pd.Series([tss_start], index=['tss_start'])\n",
    "\n",
    "#add a column consisting of gene names\n",
    "def extract_gene_name(row):\n",
    "    attributes = parse_attributes(row['attribute'])\n",
    "    gene_name = attributes.get('gene_name', '')\n",
    "    return pd.Series([gene_name], index=['gene_name'])\n",
    "\n",
    "#add gene name column\n",
    "tss_df[['gene_name']] = tss_df.apply(extract_gene_name, axis=1)\n",
    "\n",
    "#get all unique gene names\n",
    "genes = pd.unique(tss_df['gene_name'])\n",
    "print(len(genes))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>3670552</td>\n",
       "      <td>3671498</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Xkr4\"; transcript_id \"NM_001011874\"; ...</td>\n",
       "      <td>Xkr4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>4409170</td>\n",
       "      <td>4409241</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Rp1\"; transcript_id \"NM_001195662\"; g...</td>\n",
       "      <td>Rp1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>4496291</td>\n",
       "      <td>4496413</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Sox17\"; transcript_id \"NM_011441\"; ge...</td>\n",
       "      <td>Sox17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>4785573</td>\n",
       "      <td>4785726</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Mrpl15\"; transcript_id \"NM_025300\"; g...</td>\n",
       "      <td>Mrpl15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>4807893</td>\n",
       "      <td>4807982</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Lypla1\"; transcript_id \"NM_008866\"; g...</td>\n",
       "      <td>Lypla1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10176</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>63964727</td>\n",
       "      <td>63964798</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Gm13749\"; transcript_id \"NR_027824\"; ...</td>\n",
       "      <td>Gm13749</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10180</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>63984465</td>\n",
       "      <td>63984483</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Mir1958\"; transcript_id \"NR_035484\"; ...</td>\n",
       "      <td>Mir1958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10189</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>64120900</td>\n",
       "      <td>64121389</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Klf7\"; transcript_id \"NM_033563\"; gen...</td>\n",
       "      <td>Klf7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10191</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>64532804</td>\n",
       "      <td>64533020</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Creb1\"; transcript_id \"NM_133828\"; ge...</td>\n",
       "      <td>Creb1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10242</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>64606480</td>\n",
       "      <td>64608137</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Mettl21a\"; transcript_id \"NM_025964\";...</td>\n",
       "      <td>Mettl21a</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>331 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      seqname   source feature     start       end score strand frame  \\\n",
       "6        chr1  unknown    exon   3670552   3671498     .      -     .   \n",
       "25       chr1  unknown    exon   4409170   4409241     .      -     .   \n",
       "35       chr1  unknown    exon   4496291   4496413     .      -     .   \n",
       "61       chr1  unknown    exon   4785573   4785726     .      -     .   \n",
       "64       chr1  unknown    exon   4807893   4807982     .      +     .   \n",
       "...       ...      ...     ...       ...       ...   ...    ...   ...   \n",
       "10176    chr1  unknown    exon  63964727  63964798     .      +     .   \n",
       "10180    chr1  unknown    exon  63984465  63984483     .      -     .   \n",
       "10189    chr1  unknown    exon  64120900  64121389     .      -     .   \n",
       "10191    chr1  unknown    exon  64532804  64533020     .      +     .   \n",
       "10242    chr1  unknown    exon  64606480  64608137     .      -     .   \n",
       "\n",
       "                                               attribute gene_name  \n",
       "6      gene_id \"Xkr4\"; transcript_id \"NM_001011874\"; ...      Xkr4  \n",
       "25     gene_id \"Rp1\"; transcript_id \"NM_001195662\"; g...       Rp1  \n",
       "35     gene_id \"Sox17\"; transcript_id \"NM_011441\"; ge...     Sox17  \n",
       "61     gene_id \"Mrpl15\"; transcript_id \"NM_025300\"; g...    Mrpl15  \n",
       "64     gene_id \"Lypla1\"; transcript_id \"NM_008866\"; g...    Lypla1  \n",
       "...                                                  ...       ...  \n",
       "10176  gene_id \"Gm13749\"; transcript_id \"NR_027824\"; ...   Gm13749  \n",
       "10180  gene_id \"Mir1958\"; transcript_id \"NR_035484\"; ...   Mir1958  \n",
       "10189  gene_id \"Klf7\"; transcript_id \"NM_033563\"; gen...      Klf7  \n",
       "10191  gene_id \"Creb1\"; transcript_id \"NM_133828\"; ge...     Creb1  \n",
       "10242  gene_id \"Mettl21a\"; transcript_id \"NM_025964\";...  Mettl21a  \n",
       "\n",
       "[331 rows x 10 columns]"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#if the strand is -, keep the rightmost exon and if the strand is +, keep the leftmost exon\n",
    "#drop everything else\n",
    "def remove_exons(genes, tss_df):\n",
    "    for gene in genes:\n",
    "        sub_df = tss_df.loc[tss_df['gene_name']==gene]\n",
    "        dir = sub_df['strand'].iloc[0]\n",
    "        for i,row in sub_df.iterrows():\n",
    "            if i == sub_df.index[0] and dir == '+':\n",
    "                continue\n",
    "            if i == sub_df.index[-1] and dir == '-':\n",
    "                continue\n",
    "            tss_df = tss_df.drop(i)\n",
    "    return tss_df\n",
    "\n",
    "#remove exons that are not relevant to tss\n",
    "tss_df = remove_exons(genes,tss_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>tss_start</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>3670552</td>\n",
       "      <td>3671498</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Xkr4\"; transcript_id \"NM_001011874\"; ...</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>3671498</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>4409170</td>\n",
       "      <td>4409241</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Rp1\"; transcript_id \"NM_001195662\"; g...</td>\n",
       "      <td>Rp1</td>\n",
       "      <td>4409241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>4496291</td>\n",
       "      <td>4496413</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Sox17\"; transcript_id \"NM_011441\"; ge...</td>\n",
       "      <td>Sox17</td>\n",
       "      <td>4496413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>4785573</td>\n",
       "      <td>4785726</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Mrpl15\"; transcript_id \"NM_025300\"; g...</td>\n",
       "      <td>Mrpl15</td>\n",
       "      <td>4785726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>4807893</td>\n",
       "      <td>4807982</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Lypla1\"; transcript_id \"NM_008866\"; g...</td>\n",
       "      <td>Lypla1</td>\n",
       "      <td>4807893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10176</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>63964727</td>\n",
       "      <td>63964798</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Gm13749\"; transcript_id \"NR_027824\"; ...</td>\n",
       "      <td>Gm13749</td>\n",
       "      <td>63964727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10180</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>63984465</td>\n",
       "      <td>63984483</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Mir1958\"; transcript_id \"NR_035484\"; ...</td>\n",
       "      <td>Mir1958</td>\n",
       "      <td>63984483</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10189</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>64120900</td>\n",
       "      <td>64121389</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Klf7\"; transcript_id \"NM_033563\"; gen...</td>\n",
       "      <td>Klf7</td>\n",
       "      <td>64121389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10191</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>64532804</td>\n",
       "      <td>64533020</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Creb1\"; transcript_id \"NM_133828\"; ge...</td>\n",
       "      <td>Creb1</td>\n",
       "      <td>64532804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10242</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>64606480</td>\n",
       "      <td>64608137</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Mettl21a\"; transcript_id \"NM_025964\";...</td>\n",
       "      <td>Mettl21a</td>\n",
       "      <td>64608137</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>331 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      seqname   source feature     start       end score strand frame  \\\n",
       "6        chr1  unknown    exon   3670552   3671498     .      -     .   \n",
       "25       chr1  unknown    exon   4409170   4409241     .      -     .   \n",
       "35       chr1  unknown    exon   4496291   4496413     .      -     .   \n",
       "61       chr1  unknown    exon   4785573   4785726     .      -     .   \n",
       "64       chr1  unknown    exon   4807893   4807982     .      +     .   \n",
       "...       ...      ...     ...       ...       ...   ...    ...   ...   \n",
       "10176    chr1  unknown    exon  63964727  63964798     .      +     .   \n",
       "10180    chr1  unknown    exon  63984465  63984483     .      -     .   \n",
       "10189    chr1  unknown    exon  64120900  64121389     .      -     .   \n",
       "10191    chr1  unknown    exon  64532804  64533020     .      +     .   \n",
       "10242    chr1  unknown    exon  64606480  64608137     .      -     .   \n",
       "\n",
       "                                               attribute gene_name  tss_start  \n",
       "6      gene_id \"Xkr4\"; transcript_id \"NM_001011874\"; ...      Xkr4    3671498  \n",
       "25     gene_id \"Rp1\"; transcript_id \"NM_001195662\"; g...       Rp1    4409241  \n",
       "35     gene_id \"Sox17\"; transcript_id \"NM_011441\"; ge...     Sox17    4496413  \n",
       "61     gene_id \"Mrpl15\"; transcript_id \"NM_025300\"; g...    Mrpl15    4785726  \n",
       "64     gene_id \"Lypla1\"; transcript_id \"NM_008866\"; g...    Lypla1    4807893  \n",
       "...                                                  ...       ...        ...  \n",
       "10176  gene_id \"Gm13749\"; transcript_id \"NR_027824\"; ...   Gm13749   63964727  \n",
       "10180  gene_id \"Mir1958\"; transcript_id \"NR_035484\"; ...   Mir1958   63984483  \n",
       "10189  gene_id \"Klf7\"; transcript_id \"NM_033563\"; gen...      Klf7   64121389  \n",
       "10191  gene_id \"Creb1\"; transcript_id \"NM_133828\"; ge...     Creb1   64532804  \n",
       "10242  gene_id \"Mettl21a\"; transcript_id \"NM_025964\";...  Mettl21a   64608137  \n",
       "\n",
       "[331 rows x 11 columns]"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#get tss_start position based on strand direction\n",
    "tss_df['tss_start'] = tss_df.apply(extract_tss_info, axis = 1)\n",
    "tss_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>tss_start</th>\n",
       "      <th>tss_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>3670552</td>\n",
       "      <td>3671498</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Xkr4\"; transcript_id \"NM_001011874\"; ...</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>3671498</td>\n",
       "      <td>3673498</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>4409170</td>\n",
       "      <td>4409241</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Rp1\"; transcript_id \"NM_001195662\"; g...</td>\n",
       "      <td>Rp1</td>\n",
       "      <td>4409241</td>\n",
       "      <td>4411241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>4496291</td>\n",
       "      <td>4496413</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Sox17\"; transcript_id \"NM_011441\"; ge...</td>\n",
       "      <td>Sox17</td>\n",
       "      <td>4496413</td>\n",
       "      <td>4498413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>4785573</td>\n",
       "      <td>4785726</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Mrpl15\"; transcript_id \"NM_025300\"; g...</td>\n",
       "      <td>Mrpl15</td>\n",
       "      <td>4785726</td>\n",
       "      <td>4787726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>4807893</td>\n",
       "      <td>4807982</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Lypla1\"; transcript_id \"NM_008866\"; g...</td>\n",
       "      <td>Lypla1</td>\n",
       "      <td>4805893</td>\n",
       "      <td>4807893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10176</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>63964727</td>\n",
       "      <td>63964798</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Gm13749\"; transcript_id \"NR_027824\"; ...</td>\n",
       "      <td>Gm13749</td>\n",
       "      <td>63962727</td>\n",
       "      <td>63964727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10180</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>63984465</td>\n",
       "      <td>63984483</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Mir1958\"; transcript_id \"NR_035484\"; ...</td>\n",
       "      <td>Mir1958</td>\n",
       "      <td>63984483</td>\n",
       "      <td>63986483</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10189</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>64120900</td>\n",
       "      <td>64121389</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Klf7\"; transcript_id \"NM_033563\"; gen...</td>\n",
       "      <td>Klf7</td>\n",
       "      <td>64121389</td>\n",
       "      <td>64123389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10191</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>64532804</td>\n",
       "      <td>64533020</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Creb1\"; transcript_id \"NM_133828\"; ge...</td>\n",
       "      <td>Creb1</td>\n",
       "      <td>64530804</td>\n",
       "      <td>64532804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10242</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>64606480</td>\n",
       "      <td>64608137</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Mettl21a\"; transcript_id \"NM_025964\";...</td>\n",
       "      <td>Mettl21a</td>\n",
       "      <td>64608137</td>\n",
       "      <td>64610137</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>331 rows Ã— 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      seqname   source feature     start       end score strand frame  \\\n",
       "6        chr1  unknown    exon   3670552   3671498     .      -     .   \n",
       "25       chr1  unknown    exon   4409170   4409241     .      -     .   \n",
       "35       chr1  unknown    exon   4496291   4496413     .      -     .   \n",
       "61       chr1  unknown    exon   4785573   4785726     .      -     .   \n",
       "64       chr1  unknown    exon   4807893   4807982     .      +     .   \n",
       "...       ...      ...     ...       ...       ...   ...    ...   ...   \n",
       "10176    chr1  unknown    exon  63964727  63964798     .      +     .   \n",
       "10180    chr1  unknown    exon  63984465  63984483     .      -     .   \n",
       "10189    chr1  unknown    exon  64120900  64121389     .      -     .   \n",
       "10191    chr1  unknown    exon  64532804  64533020     .      +     .   \n",
       "10242    chr1  unknown    exon  64606480  64608137     .      -     .   \n",
       "\n",
       "                                               attribute gene_name  tss_start  \\\n",
       "6      gene_id \"Xkr4\"; transcript_id \"NM_001011874\"; ...      Xkr4    3671498   \n",
       "25     gene_id \"Rp1\"; transcript_id \"NM_001195662\"; g...       Rp1    4409241   \n",
       "35     gene_id \"Sox17\"; transcript_id \"NM_011441\"; ge...     Sox17    4496413   \n",
       "61     gene_id \"Mrpl15\"; transcript_id \"NM_025300\"; g...    Mrpl15    4785726   \n",
       "64     gene_id \"Lypla1\"; transcript_id \"NM_008866\"; g...    Lypla1    4805893   \n",
       "...                                                  ...       ...        ...   \n",
       "10176  gene_id \"Gm13749\"; transcript_id \"NR_027824\"; ...   Gm13749   63962727   \n",
       "10180  gene_id \"Mir1958\"; transcript_id \"NR_035484\"; ...   Mir1958   63984483   \n",
       "10189  gene_id \"Klf7\"; transcript_id \"NM_033563\"; gen...      Klf7   64121389   \n",
       "10191  gene_id \"Creb1\"; transcript_id \"NM_133828\"; ge...     Creb1   64530804   \n",
       "10242  gene_id \"Mettl21a\"; transcript_id \"NM_025964\";...  Mettl21a   64608137   \n",
       "\n",
       "        tss_end  \n",
       "6       3673498  \n",
       "25      4411241  \n",
       "35      4498413  \n",
       "61      4787726  \n",
       "64      4807893  \n",
       "...         ...  \n",
       "10176  63964727  \n",
       "10180  63986483  \n",
       "10189  64123389  \n",
       "10191  64532804  \n",
       "10242  64610137  \n",
       "\n",
       "[331 rows x 12 columns]"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Calculate TSS regions (2000 bp upstream and downstream)\n",
    "tss_df['tss_start'] = tss_df['tss_start'].astype(int)\n",
    "tss_df['tss_end'] = tss_df['tss_start']\n",
    "tss_df.loc[tss_df['strand'] == '+', 'tss_start'] = tss_df['tss_start'] - 2000\n",
    "tss_df.loc[tss_df['strand'] == '-', 'tss_end'] = tss_df['tss_end'] + 2000\n",
    "\n",
    "tss_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>tss_start</th>\n",
       "      <th>tss_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>3671498</td>\n",
       "      <td>3673498</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Xkr4\"; transcript_id \"NM_001011874\"; ...</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>3671498</td>\n",
       "      <td>3673498</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>4409241</td>\n",
       "      <td>4411241</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Rp1\"; transcript_id \"NM_001195662\"; g...</td>\n",
       "      <td>Rp1</td>\n",
       "      <td>4409241</td>\n",
       "      <td>4411241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>4496413</td>\n",
       "      <td>4498413</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Sox17\"; transcript_id \"NM_011441\"; ge...</td>\n",
       "      <td>Sox17</td>\n",
       "      <td>4496413</td>\n",
       "      <td>4498413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>4785726</td>\n",
       "      <td>4787726</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Mrpl15\"; transcript_id \"NM_025300\"; g...</td>\n",
       "      <td>Mrpl15</td>\n",
       "      <td>4785726</td>\n",
       "      <td>4787726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>4805893</td>\n",
       "      <td>4807893</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Lypla1\"; transcript_id \"NM_008866\"; g...</td>\n",
       "      <td>Lypla1</td>\n",
       "      <td>4805893</td>\n",
       "      <td>4807893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10176</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>63962727</td>\n",
       "      <td>63964727</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Gm13749\"; transcript_id \"NR_027824\"; ...</td>\n",
       "      <td>Gm13749</td>\n",
       "      <td>63962727</td>\n",
       "      <td>63964727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10180</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>63984483</td>\n",
       "      <td>63986483</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Mir1958\"; transcript_id \"NR_035484\"; ...</td>\n",
       "      <td>Mir1958</td>\n",
       "      <td>63984483</td>\n",
       "      <td>63986483</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10189</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>64121389</td>\n",
       "      <td>64123389</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Klf7\"; transcript_id \"NM_033563\"; gen...</td>\n",
       "      <td>Klf7</td>\n",
       "      <td>64121389</td>\n",
       "      <td>64123389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10191</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>64530804</td>\n",
       "      <td>64532804</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Creb1\"; transcript_id \"NM_133828\"; ge...</td>\n",
       "      <td>Creb1</td>\n",
       "      <td>64530804</td>\n",
       "      <td>64532804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10242</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>64608137</td>\n",
       "      <td>64610137</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Mettl21a\"; transcript_id \"NM_025964\";...</td>\n",
       "      <td>Mettl21a</td>\n",
       "      <td>64608137</td>\n",
       "      <td>64610137</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>331 rows Ã— 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      seqname   source feature     start       end score strand frame  \\\n",
       "6        chr1  unknown    exon   3671498   3673498     .      -     .   \n",
       "25       chr1  unknown    exon   4409241   4411241     .      -     .   \n",
       "35       chr1  unknown    exon   4496413   4498413     .      -     .   \n",
       "61       chr1  unknown    exon   4785726   4787726     .      -     .   \n",
       "64       chr1  unknown    exon   4805893   4807893     .      +     .   \n",
       "...       ...      ...     ...       ...       ...   ...    ...   ...   \n",
       "10176    chr1  unknown    exon  63962727  63964727     .      +     .   \n",
       "10180    chr1  unknown    exon  63984483  63986483     .      -     .   \n",
       "10189    chr1  unknown    exon  64121389  64123389     .      -     .   \n",
       "10191    chr1  unknown    exon  64530804  64532804     .      +     .   \n",
       "10242    chr1  unknown    exon  64608137  64610137     .      -     .   \n",
       "\n",
       "                                               attribute gene_name  tss_start  \\\n",
       "6      gene_id \"Xkr4\"; transcript_id \"NM_001011874\"; ...      Xkr4    3671498   \n",
       "25     gene_id \"Rp1\"; transcript_id \"NM_001195662\"; g...       Rp1    4409241   \n",
       "35     gene_id \"Sox17\"; transcript_id \"NM_011441\"; ge...     Sox17    4496413   \n",
       "61     gene_id \"Mrpl15\"; transcript_id \"NM_025300\"; g...    Mrpl15    4785726   \n",
       "64     gene_id \"Lypla1\"; transcript_id \"NM_008866\"; g...    Lypla1    4805893   \n",
       "...                                                  ...       ...        ...   \n",
       "10176  gene_id \"Gm13749\"; transcript_id \"NR_027824\"; ...   Gm13749   63962727   \n",
       "10180  gene_id \"Mir1958\"; transcript_id \"NR_035484\"; ...   Mir1958   63984483   \n",
       "10189  gene_id \"Klf7\"; transcript_id \"NM_033563\"; gen...      Klf7   64121389   \n",
       "10191  gene_id \"Creb1\"; transcript_id \"NM_133828\"; ge...     Creb1   64530804   \n",
       "10242  gene_id \"Mettl21a\"; transcript_id \"NM_025964\";...  Mettl21a   64608137   \n",
       "\n",
       "        tss_end  \n",
       "6       3673498  \n",
       "25      4411241  \n",
       "35      4498413  \n",
       "61      4787726  \n",
       "64      4807893  \n",
       "...         ...  \n",
       "10176  63964727  \n",
       "10180  63986483  \n",
       "10189  64123389  \n",
       "10191  64532804  \n",
       "10242  64610137  \n",
       "\n",
       "[331 rows x 12 columns]"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Ensure the start is always less than the end\n",
    "tss_df['start'] = tss_df[['tss_start', 'tss_end']].min(axis=1)\n",
    "tss_df['end'] = tss_df[['tss_start', 'tss_end']].max(axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>tss_start</th>\n",
       "      <th>tss_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>3671498</td>\n",
       "      <td>3673498</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Xkr4\"; transcript_id \"NM_001011874\"; ...</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>3671498</td>\n",
       "      <td>3673498</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>4409241</td>\n",
       "      <td>4411241</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Rp1\"; transcript_id \"NM_001195662\"; g...</td>\n",
       "      <td>Rp1</td>\n",
       "      <td>4409241</td>\n",
       "      <td>4411241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>4496413</td>\n",
       "      <td>4498413</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Sox17\"; transcript_id \"NM_011441\"; ge...</td>\n",
       "      <td>Sox17</td>\n",
       "      <td>4496413</td>\n",
       "      <td>4498413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>4785726</td>\n",
       "      <td>4787726</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Mrpl15\"; transcript_id \"NM_025300\"; g...</td>\n",
       "      <td>Mrpl15</td>\n",
       "      <td>4785726</td>\n",
       "      <td>4787726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>4805893</td>\n",
       "      <td>4807893</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Lypla1\"; transcript_id \"NM_008866\"; g...</td>\n",
       "      <td>Lypla1</td>\n",
       "      <td>4805893</td>\n",
       "      <td>4807893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10176</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>63962727</td>\n",
       "      <td>63964727</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Gm13749\"; transcript_id \"NR_027824\"; ...</td>\n",
       "      <td>Gm13749</td>\n",
       "      <td>63962727</td>\n",
       "      <td>63964727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10180</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>63984483</td>\n",
       "      <td>63986483</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Mir1958\"; transcript_id \"NR_035484\"; ...</td>\n",
       "      <td>Mir1958</td>\n",
       "      <td>63984483</td>\n",
       "      <td>63986483</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10189</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>64121389</td>\n",
       "      <td>64123389</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Klf7\"; transcript_id \"NM_033563\"; gen...</td>\n",
       "      <td>Klf7</td>\n",
       "      <td>64121389</td>\n",
       "      <td>64123389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10191</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>64530804</td>\n",
       "      <td>64532804</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Creb1\"; transcript_id \"NM_133828\"; ge...</td>\n",
       "      <td>Creb1</td>\n",
       "      <td>64530804</td>\n",
       "      <td>64532804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10242</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>64608137</td>\n",
       "      <td>64610137</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Mettl21a\"; transcript_id \"NM_025964\";...</td>\n",
       "      <td>Mettl21a</td>\n",
       "      <td>64608137</td>\n",
       "      <td>64610137</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>331 rows Ã— 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      seqname   source feature     start       end score strand frame  \\\n",
       "6        chr1  unknown    exon   3671498   3673498     .      -     .   \n",
       "25       chr1  unknown    exon   4409241   4411241     .      -     .   \n",
       "35       chr1  unknown    exon   4496413   4498413     .      -     .   \n",
       "61       chr1  unknown    exon   4785726   4787726     .      -     .   \n",
       "64       chr1  unknown    exon   4805893   4807893     .      +     .   \n",
       "...       ...      ...     ...       ...       ...   ...    ...   ...   \n",
       "10176    chr1  unknown    exon  63962727  63964727     .      +     .   \n",
       "10180    chr1  unknown    exon  63984483  63986483     .      -     .   \n",
       "10189    chr1  unknown    exon  64121389  64123389     .      -     .   \n",
       "10191    chr1  unknown    exon  64530804  64532804     .      +     .   \n",
       "10242    chr1  unknown    exon  64608137  64610137     .      -     .   \n",
       "\n",
       "                                               attribute gene_name  tss_start  \\\n",
       "6      gene_id \"Xkr4\"; transcript_id \"NM_001011874\"; ...      Xkr4    3671498   \n",
       "25     gene_id \"Rp1\"; transcript_id \"NM_001195662\"; g...       Rp1    4409241   \n",
       "35     gene_id \"Sox17\"; transcript_id \"NM_011441\"; ge...     Sox17    4496413   \n",
       "61     gene_id \"Mrpl15\"; transcript_id \"NM_025300\"; g...    Mrpl15    4785726   \n",
       "64     gene_id \"Lypla1\"; transcript_id \"NM_008866\"; g...    Lypla1    4805893   \n",
       "...                                                  ...       ...        ...   \n",
       "10176  gene_id \"Gm13749\"; transcript_id \"NR_027824\"; ...   Gm13749   63962727   \n",
       "10180  gene_id \"Mir1958\"; transcript_id \"NR_035484\"; ...   Mir1958   63984483   \n",
       "10189  gene_id \"Klf7\"; transcript_id \"NM_033563\"; gen...      Klf7   64121389   \n",
       "10191  gene_id \"Creb1\"; transcript_id \"NM_133828\"; ge...     Creb1   64530804   \n",
       "10242  gene_id \"Mettl21a\"; transcript_id \"NM_025964\";...  Mettl21a   64608137   \n",
       "\n",
       "        tss_end  \n",
       "6       3673498  \n",
       "25      4411241  \n",
       "35      4498413  \n",
       "61      4787726  \n",
       "64      4807893  \n",
       "...         ...  \n",
       "10176  63964727  \n",
       "10180  63986483  \n",
       "10189  64123389  \n",
       "10191  64532804  \n",
       "10242  64610137  \n",
       "\n",
       "[331 rows x 12 columns]"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Remove potential negative starts\n",
    "tss_df['start'] = tss_df['start'].clip(lower=0)\n",
    "\n",
    "tss_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Create BED file format\n",
    "tss_bed = tss_df[['seqname', 'start', 'end', 'gene_name', 'score', 'strand']]\n",
    "\n",
    "# Write BED file - replace with your actual file path\n",
    "tss_bed.to_csv('tss_regions.bed', sep='\\t', header=False, index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
